2_correlation_analysis/spell_vs_counts_experiment/1_correlation_analysis.py

```diff
@@ -0,0 +1,253 @@
+#
+# From [Hibbs et. al.](https://academic.oup.com/bioinformatics/article/23/20/2692/229926), we apply their "signal balancing technique that enhances biological information". This is the first part of their [SPELL](https://spell.yeastgenome.org/) algorithm that is described in section 2.3.1. SPELL calculates the correlation on the gene coefficient matrix, $U$ (i.e. how much genes contribute to a latent variable) that is generated after applying SVD. This matrix represents how genes contribute to independent latent variables that capture the signal in the data where the variance of the variables is 1. The idea is that correlations between gene contributions are more balanced so that less prominent patterns are amplified and more dominant patterns are dampended due to this compression. Figure 3 shows how well SPELL recapitulates biology (i.e. the relationship between genes within a GO term) compared to Pearson correlation.
+
+# Transpose compendia to be gene x sample
+# Here we're interested in how genes cluster
+pao1_compendium_T = pao1_compendium.T
+pa14_compendium_T = pa14_compendium.T
+
+# log transform data
+pao1_compendium_log10 = np.log10(1 + pao1_compendium_T)
+pa14_compendium_log10 = np.log10(1 + pa14_compendium_T)
+
+# Apply SVD
+pao1_U, pao1_s, pao1_Vh = np.linalg.svd(pao1_compendium_log10, full_matrices=False)
+pa14_U, pa14_s, pa14_Vh = np.linalg.svd(pa14_compendium_log10, full_matrices=False)
+
+print(pao1_compendium_T.shape)
+print(pao1_U.shape, pao1_s.shape, pao1_Vh.shape)
+
+print(pa14_compendium_T.shape)
+print(pa14_U.shape, pa14_s.shape, pa14_Vh.shape)
+
+# Convert ndarray to df to use corr()
+pao1_U_df = pd.DataFrame(data=pao1_U, index=pao1_compendium_T.index)
+pa14_U_df = pd.DataFrame(data=pa14_U, index=pa14_compendium_T.index)
+
+# Correlation of U
+# Since `corr()` computes pairwise correlation of columns we need to invert U
+pao1_corr_log_spell = pao1_U_df.iloc[:, :num_SVs].T.corr()
+pa14_corr_log_spell = pa14_U_df.iloc[:, :num_SVs].T.corr()
+
+# Check for duplicates indices
+assert pao1_corr_log_spell.index.duplicated().sum() == 0
+assert pa14_corr_log_spell.index.duplicated().sum() == 0
+
+# Check for duplicate rows
+assert pao1_corr_log_spell[pao1_corr_log_spell.duplicated(keep=False)].shape[0] == 0
+assert pa14_corr_log_spell[pa14_corr_log_spell.duplicated(keep=False)].shape[0] == 0
+
+# +
+# Plot heatmap
+h1a = sns.clustermap(pao1_corr_log_spell, cmap="BrBG", center=0, figsize=(20, 20))
+h1a.fig.suptitle(
+    f"log transform + SPELL corrected using {num_SVs} vectors (PAO1)",
+    y=1.05,
+    fontsize=24,
+)
+
+# Save
+pao1_log_spell_filename = os.path.join(
+    paths.LOCAL_DATA_DIR, f"pao1_{subset_genes}_log_spell_clustermap_test.png"
+)
+h1a.savefig(pao1_log_spell_filename, dpi=300)
+
+# +
+h2a = sns.clustermap(pa14_corr_log_spell, cmap="BrBG", center=0, figsize=(20, 20))
```

Comment:
These do look very different than the heatmaps without the correction. I suspect you'd find many more communities if you did community detection on this matrix, and the sizes would be much more uniform (not sure if this is in your plans or not, just something I'd be curious about).

Reply:
We're planning to perform a clustering analysis on the accessory-accessory correlation matrix to extract accessory modules. But for the core-core correlation matrix, we will be performing an analysis using the correlation matrix directly. Stay tuned!